designs/Transitive-Dependencies.md

```diff
@@ -0,0 +1,150 @@
+
+When a user highlights a transitive package, they will see a pop-up that displays how the transitive dependency originated & what top-level package(s) are bringing it in.
+
+![](../resources/TransitiveDependencies/TransitiveVSPMUI.png)
+
+Finally, selecting a transitive dependency will provide the user the ability to “Install” the package as you normally would a top-level dependency. This will promote the transitive dependency to a top-level dependency, and the transitive dependency will now be placed in the top-level packages list instead.
+
+## Goals and Non-Goals
+
+### Goals
+
+-	View transitive dependency list displaying the version of each dependency.
+-	See what brought in a transitive dependency & what version. 
+-	Ability to promote a transitive dependency to a top-level dependency via typical NuGet PM UI commands.
+
+### Non-Goals
+
+-	Complete management of transitive dependencies.
+-	Reinvent common UX patterns used in Visual Studio today for package management actions.
+
+## Design
+
+### Elements
+
+There are three primary elements that will be added as part of this design to the existing Visual Studio Package Manager UI.
+
+1.	ScrollViewer – To show transitive dependency nodes.
+2.	Expander – To collapse transitive dependency nodes by default & expand upon user action. To collapse top-level dependency nodes.
+3.	Tooltip – To show the transitive dependency’s origin.
+
+## Customer Development
+
+### Pain Points
+
+-	Frustration with lack of transparency for transitive packages during installs (did not realize installing package X would bring in so much other stuff)
+-	Frustration with lack of transparency for debugging reasons on PackageReference (customers feel they would be able to better debug versioning issues if they had a clearer view of their dependency graph)
+-	Some complaints about licensing issues in the case where a top-level package has a compatible license, but a transitive package does not.
+
+### Verbatims
+
+- Auditing/upgrading transitive dependencies. I can only do it manually by adding explicit packages.
+- Drilling down the dependency tree
+- In the package detail view, some way to explore dependencies of dependencies and then their dependencies and so on, any level deep, without installing the package.
+
+
+## Cross-cutting Concerns
+
+-	The words “top-level” & “transitive” are words most people must stop & think about.
+    -	Should we use “direct” & “indirect”? Or “explicit” & “implicit”?
+-	Should a transitive dependency have the “installed” text given it’s already technically installed? Should it be called “promote” instead?
+- Given that the tooltip is not accessible for .NET Framework workloads, we will have to come up with an alternative experience or cut the tooltip experience until we have an accessible experience.
```

Comment:
Can the reasons why be elaborated? It may have been mentioned in a meeting, but I can't recall and think the details should be here.

Reply:
1. .NET Core has accessibility properties for Tooltip, .NET Framework supposedly doesn't.
2. Tooltip doesn't scale well with listing potentially hundreds of packages that may bring in the transitive dependency.
3. UX Board wanted to introduce a new concept, but hasn't scheduled a workshop yet.

Reply:
Was resolved but I don't see the details added to the spec

Reply:
Moved this concern to the Elements section.